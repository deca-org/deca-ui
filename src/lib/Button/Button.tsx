import { ThemeContext } from '@lib/Theme';
import { CSS, StandardColors } from '@lib/Theme/stitches.config';
import {
  __DEV__,
  PolymorphicComponentPropWithRef,
  PolymorphicRef,
} from '@lib/Utils';
import clsx from 'clsx';
import React, { useMemo } from 'react';

import StyledButton from './Button.styles';
import ButtonIcon from './ButtonIcon';

/**
 * Buttons allow users to take actions, and make choices, with a single tap.
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * The role attribute of the component.
   */
  role?: string;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Disabled state applied to the component.
   * @default false
   */
  disabled?: boolean;
  /**
   * Function that executes when button is clicked.
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Size of the component.
   * @default md
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * The variant to use.
   * @default solid
   */
  variant?: 'solid' | 'outlined' | 'ghost' | 'flat';
  /**
   * Color to use.
   * @default primary
   */
  color?: StandardColors;
  /**
   * Set button width to 100%.
   * @default false
   */
  maxWidth?: boolean;
  /**
   * Icon on the left side of the component.
   */
  icon?: React.ReactNode;
  /**
   * Icon on the right side of the component.
   */
  iconRight?: React.ReactNode;
  /**
   * CSS applied to icon on the left side of the component.
   */
  iconLeftCss?: CSS;
  /**
   * CSS applied to icon on the right side of the component.
   */
  iconRightCss?: CSS;
  /**
   * Have button be pill shaped
   */
  pill?: boolean;
}

export type ButtonProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type ButtonComponent = (<C extends React.ElementType = 'button'>(
  props: ButtonProps<C>
) => React.ReactElement | null) & { displayName?: string };

export const Button: ButtonComponent = React.forwardRef(
  <T extends React.ElementType = 'button'>(
    {
      role = 'button',
      as,
      css,
      icon,
      iconRight,
      iconLeftCss,
      iconRightCss,
      onClick,
      color = 'primary',
      disabled = false,
      variant = 'solid',
      size = 'md',
      maxWidth = false,
      className = '',
      children,
      pill = false,
      ...btnProps
    }: ButtonProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const hasText = useMemo(
      () => React.Children.count(children) !== 0,
      [children]
    );

    // ensures only icon or only iconRight is present
    const singleIcon = useMemo(
      () =>
        (!hasText && !iconRight && icon) || (!hasText && !icon && iconRight)
          ? true
          : false,
      [children, iconRight, icon]
    );

    if (disabled) {
      onClick = undefined;
    }

    const preClass = 'decaButton';

    const { dark } = React.useContext(ThemeContext);

    return (
      <StyledButton
        role={role}
        as={as}
        css={css}
        onClick={onClick}
        color={color}
        disabled={disabled}
        isDisabled={disabled}
        variant={variant}
        size={size}
        maxWidth={maxWidth}
        className={clsx(className, `${preClass}-root`)}
        ref={ref}
        singleIcon={singleIcon}
        pill={pill}
        isDark={dark}
        {...btnProps}
      >
        {(icon && hasText) || (icon && iconRight) ? (
          <ButtonIcon
            size={size}
            isSingle={false}
            side="left"
            css={iconLeftCss}
            className={`${preClass}-leftIcon`}
          >
            {icon}
          </ButtonIcon>
        ) : (
          icon && (
            <ButtonIcon
              size={size}
              isSingle={true}
              side="left"
              css={iconLeftCss}
              className={`${preClass}-leftIcon`}
            >
              {icon}
            </ButtonIcon>
          )
        )}
        {hasText && children}
        {(iconRight && hasText) || (icon && iconRight) ? (
          <ButtonIcon
            size={size}
            isSingle={false}
            side="right"
            css={iconRightCss}
            className={`${preClass}-rightIcon`}
          >
            {iconRight}
          </ButtonIcon>
        ) : (
          iconRight && (
            <ButtonIcon
              size={size}
              isSingle={true}
              side="right"
              css={iconRightCss}
              className={`${preClass}-rightIcon`}
            >
              {iconRight}
            </ButtonIcon>
          )
        )}
      </StyledButton>
    );
  }
);

if (__DEV__) {
  Button.displayName = 'DecaUI.Button';
}

export default Button;
