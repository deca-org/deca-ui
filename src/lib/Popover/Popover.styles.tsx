import { styled } from '@lib/Theme/stitches.config';
import { animated } from '@react-spring/web';

export const StyledPopover = styled(animated.div, {
  fontFamily: '$normal',
  p: '$3',
  boxShadow: '$default',
  br: '$sm',
  color: '$text',
  zIndex: '$5',
  variants: {
    isDark: {
      true: {
        bg: '$popperDarkBg',
      },
      false: {
        bg: '$popperLightBg',
      },
    },
  },
});
