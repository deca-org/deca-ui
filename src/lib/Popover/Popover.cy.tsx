import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Popover, { PopoverProps } from '.';
import Button from '../Button';

const PopoverComposer = (props: PopoverProps) => (
  <Popover {...props}>
    <Popover.Trigger>
      <Button>Open Popover</Button>
    </Popover.Trigger>
    <Popover.Content>This is the content of the popover.</Popover.Content>
  </Popover>
);

describe('components/Popover', () => {
  describe('base', () => {
    it('background-color', () => {
      cy.mount(<PopoverComposer />);
      cy.get('button').click();
      cy.get('.decaPopover-root').should(
        'have.css',
        'background-color',
        Test.color('popperLightBg')
      );
    });
    it('color', () => {
      cy.mount(<PopoverComposer />);
      cy.get('button').click();
      cy.get('.decaPopover-root').should(
        'have.css',
        'color',
        Test.color('black')
      );
    });
  });
  describe('state', () => {
    it('not exist in DOM when not open', () => {
      cy.mount(<PopoverComposer />);
      cy.get('.decaPopover-root').should('not.exist');
    });
    it('exists in DOM when open (button clicked)', () => {
      cy.mount(<PopoverComposer />);
      cy.get('button').click();
      cy.get('.decaPopover-root').should('exist');
    });
  });
  describe('dark mode', () => {
    it('background-color', () => {
      cy.darkMount(<PopoverComposer />);
      cy.get('button').click();
      cy.get('.decaPopover-root').should(
        'have.css',
        'background-color',
        Test.color('popperDarkBg')
      );
    });
    it('color', () => {
      cy.darkMount(<PopoverComposer />);
      cy.get('button').click();
      cy.get('.decaPopover-root').should(
        'have.css',
        'color',
        Test.color('white')
      );
    });
  });
});
