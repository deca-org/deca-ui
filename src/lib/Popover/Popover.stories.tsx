import Button from '@lib/Button';
import Popover from '@lib/Popover';
import { DecaUIProvider } from '@lib/Theme';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

export default {
  title: 'Popover',
  component: Popover,
} as ComponentMeta<typeof Popover>;

const Template: ComponentStory<typeof Popover> = (args) => (
  <Popover {...args}>
    <Popover.Trigger>
      <Button>Open Popover</Button>
    </Popover.Trigger>
    <Popover.Content>This is the content of the popover.</Popover.Content>
  </Popover>
);

export const Default = Template.bind({});
Default.args = {
  placement: 'bottom',
  action: 'click',
};

export const WithTheme = Template.bind({});

WithTheme.args = { ...Default.args };
WithTheme.decorators = [
  (Story) => (
    <DecaUIProvider
      theme={{
        colors: {
          primary: '$green600',
        },
        radii: {
          sm: '15px',
        },
      }}
    >
      <Story />
    </DecaUIProvider>
  ),
];

export const DarkMode = Template.bind({});

DarkMode.args = { ...Default.args };
DarkMode.parameters = { backgrounds: { default: 'dark' } };

DarkMode.decorators = [
  (Story) => (
    <DecaUIProvider mode="dark">
      <Story />
    </DecaUIProvider>
  ),
];
