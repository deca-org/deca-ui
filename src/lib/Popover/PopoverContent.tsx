import { ThemeContext } from '@lib/Theme';
import { CSS } from '@lib/Theme/stitches.config';
import {
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  mergeRefs,
  useClickOutside,
  __DEV__,
} from '@lib/Utils';
import { animated, useTransition } from '@react-spring/web';
import clsx from 'clsx';
import React, { useState, useEffect, useContext } from 'react';
import ReactDOM from 'react-dom';

import { PopoverContext, IPopoverContext } from './Popover';
import { StyledPopover } from './Popover.styles';

/**
 * PopoverContent contains the content shown when the trigger is executed
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
}

export type PopoverContentProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type PopoverContentComponent = (<C extends React.ElementType = 'div'>(
  props: PopoverContentProps<C>
) => React.ReactElement | null) & { displayName?: string };

const PopoverContent: PopoverContentComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    { children, css, className = '', as, ...props }: PopoverContentProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const context = useContext(PopoverContext) as IPopoverContext;

    const clickOutsideRef = useClickOutside(() => {
      context.setOpen && context.setOpen(false);
    }, [context.triggerRef]);

    const transition = useTransition(context.open, {
      from: {
        scale: 0.75,
        opacity: 0,
      },
      enter: {
        scale: 1,
        opacity: 1,
      },
      leave: {
        scale: 0.75,
        opacity: 0,
      },
      config: {
        tension: 300,
        friction: 19,
      },
    });

    const preClass = 'decaPopover';

    const { dark } = React.useContext(ThemeContext);

    const [DOM, setDOM] = useState(false);

    useEffect(() => {
      setDOM(true);
    }, []);

    if (DOM) {
      return ReactDOM.createPortal(
        transition(
          (style, item) =>
            item && (
              <StyledPopover
                style={style}
                ref={mergeRefs(
                  context.mainComponentRef,
                  context.floating,
                  clickOutsideRef,
                  ref
                )}
                css={{
                  position: context.strategy,
                  top: context.y ?? 0,
                  left: context.x ?? 0,
                  ...css,
                }}
                className={clsx(className, `${preClass}-root`)}
                as={animated[as as keyof JSX.IntrinsicElements]}
                isDark={dark}
                {...props}
              >
                {children}
              </StyledPopover>
            )
        ),
        document.getElementById('decaUI-provider')
          ? (document.getElementById('decaUI-provider') as Element)
          : (document.querySelector('body') as Element)
      );
    }
    return <></>;
  }
);

if (__DEV__) {
  PopoverContent.displayName = 'DecaUI.PopoverContent';
}

export default PopoverContent;
