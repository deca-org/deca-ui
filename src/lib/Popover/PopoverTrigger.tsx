import { mergeRefs, __DEV__ } from '@lib/Utils';
import React, { useContext } from 'react';

import { PopoverContext, IPopoverContext } from './Popover';

/**
 * PopoverTrigger opens the popover's content. It must be an interactive element
 * such as `button` or `a`.
 */
export interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
}

const PopoverTrigger = ({ children }: Props) => {
  const context = useContext(PopoverContext) as IPopoverContext;

  // enforce single child
  const child: any = React.Children.only(children);

  if (context.action === 'click') {
    const extendedOnClick = () => {
      context.setOpen && context.setOpen((prevState) => !prevState);
      child.props.onClick && child.props.onClick();
    };

    return React.cloneElement(child, {
      ...child.props,
      onClick: extendedOnClick,
      ref: mergeRefs(context.reference, child.ref, context.triggerRef),
    });
  } else {
    const extendedOnMouseEnter = () => {
      context.setOpen && context.setOpen(true);
      child.props.onMouseEnter && child.props.onMouseEnter();
    };

    const extendedOnMouseLeave = () => {
      context.setOpen && context.setOpen(false);
      child.props.onMouseLeave && child.props.onMouseLeave();
    };

    return React.cloneElement(child, {
      ...child.props,
      onMouseEnter: extendedOnMouseEnter,
      onMouseLeave: extendedOnMouseLeave,
      ref: mergeRefs(context.reference, child.ref, context.triggerRef),
    });
  }
};

if (__DEV__) {
  PopoverTrigger.displayName = 'DecaUI.PopoverTrigger';
}

export default PopoverTrigger;
