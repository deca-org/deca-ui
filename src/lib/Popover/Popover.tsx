import {
  useFloating,
  offset as floatingOffset,
  flip,
  shift,
  autoUpdate,
  UseFloatingReturn,
  Placement,
} from '@floating-ui/react-dom';
import {
  MasterComponent,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  __DEV__,
} from '@lib/Utils';
import React, {
  useState,
  useEffect,
  useMemo,
  SetStateAction,
  Dispatch,
} from 'react';
import PopoverTrigger from './PopoverTrigger';
import PopoverContent from './PopoverContent';

/**
 * A Popover can be used to display some content on top of another.
 */
export interface Props {
  /**
   * The content of the component. It is usually the `Popover.Trigger`,
   * and `Popover.Content`
   */
  children?: React.ReactNode[];
  /*
   * If true, the component is shown.
   */
  open?: boolean;
  /**
   * State dispatcher function (setter in useState)
   */
  setOpen?: Dispatch<SetStateAction<boolean>>;
  /**
   * Placement of the popover component
   * @default bottom
   */
  placement?: Placement;
  /**
   * Determines what action needs to take place in order for popover to appear
   * @default click
   */
  action?: 'click' | 'hover';
  /**
   * How far away PopoverContent should be away from PopoverTrigger when opened
   * @default 10
   */
  offset?: number;
}

export interface IPopoverContext extends UseFloatingReturn {
  triggerRef?: React.Ref<HTMLElement | undefined>;
  open?: boolean;
  setOpen?: Dispatch<SetStateAction<boolean>>;
  mainComponentRef: PolymorphicRef<React.ElementType>;
  action: 'click' | 'hover';
}

export const PopoverContext = React.createContext<IPopoverContext | null>(null);

export type PopoverProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type PopoverComponent = (<C extends React.ElementType = 'div'>(
  props: PopoverProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Popover: PopoverComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      children,
      open,
      setOpen,
      placement = 'bottom',
      action = 'click',
      offset = 10,
    }: PopoverProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const floatingProps = useFloating({
      placement: placement,
      whileElementsMounted: autoUpdate,
      strategy: 'absolute',
      middleware: [floatingOffset(offset), flip(), shift()],
    });

    const [selfOpen, setSelfOpen] = useState<boolean>(false);

    const [scrollPos, setScrollPos] = useState(0);

    const isControlledComponent = useMemo(() => open !== undefined, [open]);

    const isScrolling = () => {
      if (window.scrollY !== scrollPos) {
        isControlledComponent ? setOpen && setOpen(false) : setSelfOpen(false);
        setScrollPos(window.scrollY);
      }
    };

    const handleEsc = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        isControlledComponent ? setOpen && setOpen(false) : setSelfOpen(false);
      }
    };

    useEffect(() => {
      window.addEventListener('scroll', isScrolling);
      window.addEventListener('keydown', handleEsc);
      return () => {
        window.removeEventListener('scroll', isScrolling);
        window.removeEventListener('keydown', handleEsc);
      };
    }, []);

    const [trigger, content] = React.Children.toArray(children);

    const triggerRef = React.useRef();

    return (
      <PopoverContext.Provider
        value={{
          ...floatingProps,
          triggerRef: triggerRef,
          open: isControlledComponent ? open : selfOpen,
          setOpen: isControlledComponent ? setOpen : setSelfOpen,
          mainComponentRef: ref,
          action,
        }}
      >
        {trigger}
        {content}
      </PopoverContext.Provider>
    );
  }
);

if (__DEV__) {
  Popover.displayName = 'DecaUI.Popover';
}

export default Popover as MasterComponent<
  HTMLDivElement,
  PopoverProps<React.ElementType>,
  {
    Trigger: typeof PopoverTrigger;
    Content: typeof PopoverContent;
  }
>;
