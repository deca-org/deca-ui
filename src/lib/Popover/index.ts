import Popover from './Popover';
import PopoverContent from './PopoverContent';
import PopoverTrigger from './PopoverTrigger';

Popover.Trigger = PopoverTrigger;
Popover.Content = PopoverContent;

export * from './Popover';
export { Popover };
export default Popover;
