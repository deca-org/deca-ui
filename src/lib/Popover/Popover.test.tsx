import Button from '@lib/Button';
import Popover from '@lib/Popover';
import { render, fireEvent, act, screen } from '@testing-library/react';
import React from 'react';

describe('components/Popover', () => {
  global.ResizeObserver = require('resize-observer-polyfill');
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Popover>
        <Popover.Trigger>
          <Button>Open Popover</Button>
        </Popover.Trigger>
        <Popover.Content>content</Popover.Content>
      </Popover>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('click action', () => {
    const utils = render(
      <Popover>
        <Popover.Trigger>
          <Button>Open Popover</Button>
        </Popover.Trigger>
        <Popover.Content>content</Popover.Content>
      </Popover>
    );
    expect(screen.queryByText('content')).toBe(null);
    const popoverTrigger = utils.getByText('Open Popover');
    act(() => {
      popoverTrigger.click();
    });
    expect(screen.queryByText('content')).not.toBe(null);
  });
  it('hover action', async () => {
    const utils = render(
      <Popover action="hover">
        <Popover.Trigger>
          <Button>Open Popover</Button>
        </Popover.Trigger>
        <Popover.Content>content</Popover.Content>
      </Popover>
    );
    expect(screen.queryByText('content')).toBe(null);
    const popoverTrigger = utils.getByText('Open Popover');
    act(() => {
      fireEvent.mouseEnter(popoverTrigger);
    });
    expect(screen.queryByText('content')).not.toBe(null);
    act(() => {
      fireEvent.mouseLeave(popoverTrigger);
    });
    await act(async () => new Promise((r) => setTimeout(r, 1000)));
    expect(screen.queryByText('content')).toBe(null);
  });
});
