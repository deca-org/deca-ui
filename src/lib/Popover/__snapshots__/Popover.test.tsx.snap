// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Popover matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="c-hpqThE c-hpqThE-jsOpSy-color-primary c-hpqThE-dcPSsq-size-md c-hpqThE-bDGmTT-isDisabled-false c-hpqThE-cycqzJ-pill-false c-hpqThE-ePLPad-cv decaButton-root"
    role="button"
  >
    Open Popover
  </button>
</DocumentFragment>
`;
